### ISSUE TITLE
DISSECT_IMAGE_PIN_PARTITION_DEVICES leads to usage of /proc/self/fd in /proc/mounts and /proc/self/mountinfo

### ISSUE THREAD

**Issue by daandemeyer** [systemd/systemd#28831]
### systemd version the issue has been seen with

main

### Used distribution

Fedora 38

### Linux kernel version used

6.4.6-200.fc38.x86_64

### CPU architectures issue was seen on

x86_64

### Expected behaviour you didn't see

/proc/mounts and /proc/self/mountinfo contain the device paths after mounting an image using systemd-dissect (e.g. /dev/loop0p1 or similar)

### Unexpected behaviour you saw

/proc/mounts and /proc/self/mountinfo contain paths such as /proc/self/fd/xxx (or /proc/pid/xxx after we land https://github.com/systemd/systemd/pull/28829). This means that tooling can't reliably figure out the backing block device from /proc/mounts or /proc/self/mountinfo anymore when using `systemd-dissect --mount`, which used to be possible before we started mounting from file descriptors (which was changed by https://github.com/systemd/systemd/pull/24625. 

cc @yuwata 

### Steps to reproduce the problem

`systemd-dissect --mount "image" /abc` followed by `cat /proc/self/mountinfo`

**Comment by poettering** [comment]
yeah, this sucks, but my assumption was that for DDIs that doesn't matter too much, because usually they are mounted in some namespace, not the host.

@brauner has been looking into extending the new mount API so that we can just pass in the fd to the block device directly to the mount requests, instead of going via /proc/self/fd/, but until then not sure we can do about this. 

THis is ultimately just cosmetics I guess...

We could do a terrible hack like this:

1. create a temporary directory
2. chdir() into the dir
3. create symlinks in it pointing to /proc/self/fd/
4. mount by specifiying the symlink as source, as a relative path (i.e. just the filename).
5. remove symlink and temporary directory again

This would give us the ability to name the source any way we want, with some restrictions, such as no starting "/"...

**Comment by poettering** [comment]
another terrible hack would be: open both source device node, and target inode as fd. then fork of short-lived child process. detach fs namespaces there, then mount a new tmpfs to / in it, with /proc set up. Then create path /dev/foo/bar of your choice in there, and make the final part of it a symlink to /proc/self/fd/...

Then issue mount() inside that environment with the source that symlink and the target the target inode referenced directly via /proc/self/fd/…

That would give you full freedom to name the source...
